tests/test_long_insert.py

```diff
@@ -0,0 +1,19 @@
+from tests.testcase import BaseTestCase
+
+
+class LongInsertTestCase(BaseTestCase):
+    client_kwargs = {
```

Comment:
I would say that this settings deserves some comments:
- insert_block_size=1 - to produce as much blocks as possible
- send_timeout/receive_timeout - to trigger timeout in case of client does not consumes profile events packets

Reply:
@insomnes we have time limit of 10s for each test. Is 5s enough for reproducing? It seems that `pytest-timer` applies timeout to all test without any exceptions.

Should we patch `connection.receive_profile_events` or `read_varint` with following mock to ensure this test works?

```python
        def side_receive_profile_events(*args, **kwargs):
            # do something, toggle flag, etc.
            return receive_profile_events(*args, **kwargs)

        with patch('socket.getaddrinfo') as side_receive_profile_events:
            receive_profile_events.side_effect = side_receive_profile_events
            self.client.execute('INSERT INTO ...')
           # assert something happened in side_receive_profile_events
```

See `tests.test_connect.ConnectTestCase.test_alt_hosts` for inspiration.

Reply:
@xzkostyan thank you for the clarification. I would need a bit of time to wrap my head around this. This test depends on server behavior and I can't come up with an appropriate mock from the get-go.

With the unpatched function and this synthetic data, this will for sure error in the first 5 seconds as `ServerException` if we were inserting quickly enough or as our socket's `ConnectionResetError`.

With the patched one it would take a long for the proper ending cause we process 100 000 ProfileEvents packets on new server versions.
Maybe a mock that passes if we don't catch `ServerException` or `ConnectionResetError` in the first 7 seconds would be ok. 

I will try to think about it a bit more and come back with a solution.

Reply:
@xzkostyan sorry for the long pause.
I've tried to emulate the situation of bug with adding of small sleep before calling `receive_end_of_query` in `Client`.
And lowered the row count. This configuration still reproduces the bug on unpatched version, but runs normally and under 7-8 seconds in patched one.

What do you think?

Reply:
Just out of curiosity I looked at this change, and this hack looks nice, though it was not clear how it works at first glance (when the driver do not receive profile events it will call receive_end_of_query, which will read all ProfileEvents and also EndOfStream, but it will sleep each time to 2 seconds, so eventually the server will timed out, and it does not timed out for EndOfStream, because it will be written to socket buffer)

But I think that it should be fixed differently, instead `receive_end_of_query` should not handle `ProfileEvents` for `INSERT` (they still should be handled for SELECT though), so I think there should be separate `receive_end_of_insert_query` that will not handle them, and then it will be enough to send one block in this test, without any timeouts trickery.

And this will match with how ClickHouse handle this protocol internally for INSERT, so this is preferable I would say.

Reply:
Something like this - https://gist.github.com/azat/da85ff0bde3f0da259144b0ba361cd64

BTW it also founds one problem in this patch - missing reading of ProfileEvents after empty block

Reply:
@xzkostyan hello again! 

I've applied @azat suggestion with a slight adjustment of moving method to client, and accepting possible Progress event in it (I found the possibility of Progress packet coming before ProfileEvents in test). Azat confirmed that it is normal and expected behavior from the server side. 

There is only a simple insert test now. I have deleted the previous one, cause it still fails due to timeout inside github.

If you think, that some more direct test should be applied, maybe we can check that `receive_end_of_query` is not called and `receive_end_of_insert_query` has not get any `ProfileEvents` packet in it.

This breaks backward compatibility in some sens, so If you think this should be made other way, I would be happy to change code according to your vision.